zivid_camera/src/zivid_camera.cpp

```diff
@@ -93,440 +206,534 @@ Zivid::Application makeZividApplication()
-  , header_seq_(0)
-{
-  ROS_INFO("Zivid ROS driver version %s", ZIVID_ROS_DRIVER_VERSION);
-
-  ROS_INFO("Node's namespace is '%s'", nh_.getNamespace().c_str());
-  if (nh_.getNamespace() == "")
+
+namespace ParamNames
+{
+constexpr auto serial_number = "serial_number";
+constexpr auto file_camera_path = "file_camera_path";
+constexpr auto frame_id = "frame_id";
+}  // namespace ParamNames
+
+template <typename SettingsType>
+class CaptureSettingsController
+{
+public:
+  explicit CaptureSettingsController(rclcpp::Node & node)
+  : node_(node),
+    file_path_param_{baseName() + std::string{"_file_path"}},
+    yaml_string_param_{baseName() + std::string{"_yaml"}}
+  {
+    for (const auto & param : {yaml_string_param_, file_path_param_}) {
+      node_.declare_parameter<std::string>(param, "");
+    }
+  }
+
+  SettingsType currentSettings() const
   {
-    // Require the user to specify the namespace that this node will run in.
-    // See REP-135 http://www.ros.org/reps/rep-0135.html
-    throw std::runtime_error("Zivid ROS driver started in the global namespace ('/')! This is unsupported. "
-                             "Please specify namespace, e.g. using the ROS_NAMESPACE environment variable.");
+    if (cached_settings_.has_value()) {
+      RCLCPP_DEBUG_STREAM(node_.get_logger(), "Using cached settings");
+      return *cached_settings_;
+    }
+
+    const auto settings_file_path = node_.get_parameter(file_path_param_).as_string();
+    const auto settings_yaml = node_.get_parameter(yaml_string_param_).as_string();
+
+    if (!settings_file_path.empty() && !settings_yaml.empty()) {
+      logErrorToLoggerAndThrowRuntimeException(
+        node_.get_logger(),
+        "Both '" + file_path_param_ + "' and '" + yaml_string_param_ +
+          "' parameters are non-empty! Please set only one of these parameters.");
+    } else if (settings_file_path.empty() && settings_yaml.empty()) {
+      logErrorToLoggerAndThrowRuntimeException(
+        node_.get_logger(), "Both '" + file_path_param_ + "' and '" + yaml_string_param_ +
+                              "' parameters are empty! Please set one of these parameters.");
+    }
+
+    if (!settings_yaml.empty()) {
+      RCLCPP_DEBUG_STREAM(node_.get_logger(), "Using settings from yml string");
+      cached_settings_ = deserializeZividDataModel<SettingsType>(settings_yaml);
+    } else {
+      RCLCPP_DEBUG_STREAM(
+        node_.get_logger(), "Using settings from file '" << settings_file_path << "'");
+      cached_settings_ = SettingsType{settings_file_path};
+    }
+
+    return *cached_settings_;
+  }
+
+  void setSettings(const SettingsType & settings)
+  {
+    RCLCPP_INFO_STREAM(node_.get_logger(), "Setting settings from " << settings.name << " object");
+    cached_settings_.reset();
+    node_.set_parameter(rclcpp::Parameter{yaml_string_param_, serializeZividDataModel(settings)});
+    node_.set_parameter(rclcpp::Parameter{file_path_param_, ""});
+    cached_settings_ = settings;
```

Comment:
It may be this line (and `cached_settings_.reset();` above) is not needed? Will the onSetParameter() line below be called when node_.set_parameter() is called? If so, then it may be best to remove these cached_settings lines, and let the regular mechanism deal with the caching.

Reply:
Yeah, that's true. I was thinking the reset might be needed in case of exceptions in `serializeZividDataModel`, but that's not really a case we want to handle, so better let it propagate out.

And the last assignment was mostly because it seemed more efficient than interpreting the string. But this isn't really a case where we need to optimize. So I removed both of these. Also fixed up some debug logging.

56ce1a0a6e312f046e94aa23ae98be8c949ad6ac